### PR TITLE
[TASK] Deprecate plugin content element and plugin subtypes (list_type)

### PR DESCRIPTION
**Pull request by linawolf**
Depends on https://github.com/TYPO3-Documentation/TYPO3CMS-Reference-TCA/pull/1183

Resolves: https://github.com/TYPO3-Documentation/Changelog-To-Doc/issues/1071

Releases: main

### CODE REVIEW

File: Documentation/ApiOverview/ContentElements/ContentElementsWizard.rst [hunk 1]
```diff
@@ -64,7 +64,7 @@ the available groups.
 Plain content elements or plugins
 =================================
 
-You can add a content element or plain plugin (non-extbase) using method
+You can add a content element or plain plugin (non-Extbase) using method
```
Comment (sfroemkenjw) [inline Documentation/ApiOverview/ContentElements/ContentElementsWizard.rst:67]: For me it look ugly. What about:

```suggestion
You can add a content element or plain plugin (no Extbase) using method
```

File: Documentation/ApiOverview/ContentElements/ContentElementsWizard.rst [hunk 2]
```diff
@@ -74,11 +74,12 @@ of class :php:`\TYPO3\CMS\Core\Utility\ExtensionManagementUtility`.
...
 
-When you are using `CType` (recommended) for parameter `$type` the content
+When you are using `CType` for parameter `$type` the content
 element is added to the select item list of column `CType` in table `tt_content`.
 
-If you are using the default `list_type` for the parameter it is added as
-subtype.
+..  deprecated:: 13.4
+    Using the default `list_type` for the parameter is deprecated. All content
+    elements and plugins should be added as sing `CType` for parameter `$type`.
```
Comment (sfroemkenjw) [inline Documentation/ApiOverview/ContentElements/ContentElementsWizard.rst:82]: `sing`? La la lala la laaaa `string`? `single`?

File: Documentation/ApiOverview/ContentElements/CustomBackendPreview.rst [hunk 3]
```diff
@@ -118,23 +118,12 @@ approaches:
     This specifies the preview renderer only for records of type :php:`$type` as
     determined by the :ref:`type field <t3tca:types>` of your table.
 
-#.  Table and field have a :php:`subtype_value_field` TCA setting
+..  deprecated:: 13.4
+    Registration of subtypes has been deprecated.  Registration of custom
```
Comment (sfroemkenjw) [inline Documentation/ApiOverview/ContentElements/CustomBackendPreview.rst:122]:
```suggestion
    Registration of subtypes has been deprecated. Registration of custom
```

File: Documentation/ApiOverview/ContentElements/Index.rst [hunk 4]
```diff
@@ -183,12 +185,9 @@ To register such a plugin as content element you can use function
...
     :caption: EXT:my_extension/Configuration/TCA/Overrides/tt_content.php
 
-By using `'list_type'` as second parameter is is also possible to add the plugin
-as a list type. See :ref:`plugins-list_type`.
-
-**Plugins** are a specific type of content elements. Plugins use the CType='list'.
-Each plugin has its own plugin type, which is used in the database field
-tt_content.list_type. The list_type could be understood as subtype of CType.
+..  deprecated:: 13.4
+    Setting the second parameter to any value but `CType`
```
Comment (sfroemkenjw) [inline Documentation/ApiOverview/ContentElements/Index.rst:189]: Only `list_type` is deprecated. You still can call that method without any deprecations for any TCA select type. Sure, it does not make sense, but it's possible. `addPlugin` looks a little bit like `addTcaSelectItem`, it just adds a new entry to a select type.

File: Documentation/ApiOverview/ContentElements/MigrationListType.rst [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+Several steps are important in the migration from `list_type` plugins to `CType`
+plugins:
+
+*   Register plugins using the `CType` record type
+*   Create update wizard which extends :php:`\TYPO3\CMS\Install\Updates\AbstractListTypeToCTypeUpdate`
+    and add :php:`list_type` to :php:`CType` mapping for each plugin to migrate.
+*   Migrate possible FlexForm registration and add dedicated :php:`showitem` TCA
+    configuration
+*   Migrate possible PreviewRenderer registration in TCA
```
Comment (sfroemkenjw) [inline Documentation/ApiOverview/ContentElements/MigrationListType.rst:22]:
```suggestion
*   Migrate possible `PreviewRenderer` registration in TCA
```

File: Documentation/ApiOverview/ContentElements/MigrationListType.rst [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+*   Adapt possible content element wizard items in Page TSConfig, where
```
Comment (sfroemkenjw) [inline Documentation/ApiOverview/ContentElements/MigrationListType.rst:23]: I can live with `Page TSConfig`, but in our Spelling Doc we use:

```suggestion
*   Adapt possible content element wizard items in page TSconfig, where
```

File: Documentation/ApiOverview/ContentElements/MigrationListType.rst [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    :php:`list_type` is used
+*   Adapt possible content element restrictions in backend layouts or container
+    elements defined by third-party extensions like ext:content_defender
```
Comment (sfroemkenjw) [inline Documentation/ApiOverview/ContentElements/MigrationListType.rst:26]:
```suggestion
    elements defined by third-party extensions like :t3ext:content_defender
```

File: Documentation/ApiOverview/ContentElements/MigrationListType.rst [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+..  contents:: Table of content
+
+..  _plugins-list-type-migration-extbase:
+
+Migration example: Extbase plugin
+=================================
+
+..  _plugins-list-type-migration-extbase-configuration:
+
+1. Adjust the Extbase plugin configuration
```
Comment (sfroemkenjw) [inline Documentation/ApiOverview/ContentElements/MigrationListType.rst:37]: What about using: https://docs.typo3.org/m/typo3/docs-how-to-document/main/en-us/Howto/WritingReST/CheatSheet.html#ref-styled-numbered-lists-big-nums

File: Documentation/ApiOverview/ContentElements/MigrationListType.rst [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+..  _plugins-list-type-migration-extbase-flexform:
+
+2. Adjust the registration of FlexForms and additional fields
+-------------------------------------------------------------
+
+..  literalinclude:: _Migration/_tca_registration.php.diff
+    :caption: EXT:examples/Configuration/TCA/Overrides/tt_content.php (difference)
+
+The `CType` based plugin does not inherit the standard fields provided by the
+TCA of the content-element "List". These where in many cases removed by
```
Comment (sfroemkenjw) [inline Documentation/ApiOverview/ContentElements/MigrationListType.rst:67]: Nearly everywhere above you write it without `-`

```suggestion
TCA of the content element "List". These where in many cases removed by
```

File: Documentation/ApiOverview/ContentElements/MigrationListType.rst [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+plugin manually. This step is similar to adjusting the TypoScript of a
+Core-based plugin.
+
+..  _plugins-list-type-migration-extbase-flexform:
+
+2. Adjust the registration of FlexForms and additional fields
+-------------------------------------------------------------
+
+..  literalinclude:: _Migration/_tca_registration.php.diff
+    :caption: EXT:examples/Configuration/TCA/Overrides/tt_content.php (difference)
```
Comment (sfroemkenjw) [inline Documentation/ApiOverview/ContentElements/MigrationListType.rst:64]: You are writing `Line 15` below. So activating line numbers would be perfect

```suggestion
    :caption: EXT:examples/Configuration/TCA/Overrides/tt_content.php (difference)
    :linenos:
```

File: Documentation/ApiOverview/ContentElements/MigrationListType.rst [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+The `CType` based plugin does not inherit the standard fields provided by the
+TCA of the content-element "List". These where in many cases removed by
+using :confval:`subtypes_excludelist <t3tca:types-subtypes-excludelist>`.
+
+As these fields are not displayed automatically anymore you can remove this
+definition without replacement: Line 15 in the diff. If they have not been
+removed and still needed you need to add them manually to your plugins type.
```
Comment (sfroemkenjw) [inline Documentation/ApiOverview/ContentElements/MigrationListType.rst:72]:
```suggestion
removed and are still needed, you will need to manually add them to your plugin type.
```

File: Documentation/ApiOverview/ContentElements/MigrationListType.rst [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+
+..  _plugins-list-type-migration-core:
+
+Migration example: Core-based plugin
+====================================
+
+..  _plugins-list-type-migration-core-registration:
+
+1.  Adjust the plugin registration
+==================================
```
Comment (sfroemkenjw) [inline Documentation/ApiOverview/ContentElements/MigrationListType.rst:120]: Sorry, not rendered until now. No link, no content, ... look ugly

File: Documentation/ApiOverview/ContentElements/_Migration/_ext_localconf.php.diff [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+
+ ExtensionUtility::configurePlugin(
+     'Examples',
+     'FalExamples',
+        'HtmlParser',
+        [
+            HtmlParserController::class => 'index',
+        ],
++       [],
++    ExtensionUtility::PLUGIN_TYPE_CONTENT_ELEMENT,
```
Comment (sfroemkenjw) [inline Documentation/ApiOverview/ContentElements/_Migration/_ext_localconf.php.diff:12]: Something is completely broken here. 3rd Argument is an Array, but here it is "HtmlParser" and the indents of 6th (should be 5th) argument is missing 3 spaces. And in general the indents in PHP are 4 and not 7.

File: Documentation/CodeSnippets/FlexForms/Examples/PluginHaikuListRegistration.rst.txt [hunk 14]
```diff
@@ -19,21 +20,30 @@
 
     defined('TYPO3') or die();
 
+    $pluginSignature = 'examples_haiku_list';
```
Comment (sfroemkenjw) [inline Documentation/CodeSnippets/FlexForms/Examples/PluginHaikuListRegistration.rst.txt:23]: Plugin Signature is extkey without underscore + `_` + plugin name without underscore. So, it should be `examples_haikulist` or `exampleshaiku_list`.

File: Documentation/CodeSnippets/FlexForms/Examples/PluginHaikuListRegistration.rst.txt [hunk 15]
```diff
@@ -19,21 +20,30 @@
...
+        'CType',
         'examples',
     );
 
-    $GLOBALS['TCA']['tt_content']['types']['list']['subtypes_excludelist']['examples_haiku_list'] = 'pages,layout,select_key,recursive';
-
-    $GLOBALS['TCA']['tt_content']['types']['list']['subtypes_addlist']['examples_haiku_list'] = 'pi_flexform';
+    ExtensionManagementUtility::addToAllTCAtypes(
+        'tt_content',
+        '--div--;Configuration,pi_flexform,',
```
Comment (sfroemkenjw) [inline Documentation/CodeSnippets/FlexForms/Examples/PluginHaikuListRegistration.rst.txt:40]:
```suggestion
        '--div--;Configuration,pi_flexform',
```

File: Documentation/CodeSnippets/FlexForms/Examples/PluginHaikuListRegistration.rst.txt [hunk 16]
```diff
@@ -19,21 +20,30 @@
...
+        $pluginSignature,
+        'after:subheader',
```
Comment (sfroemkenjw) [inline Documentation/CodeSnippets/FlexForms/Examples/PluginHaikuListRegistration.rst.txt:42]: `subheader` is only available for CType `header`. As subheader is not available for all other CTypes this field will be added to the last tab and behind last field.

File: Documentation/ExtensionArchitecture/BestPractises/NamingConventions.rst [hunk 17]
```diff
@@ -365,45 +370,49 @@ Example:
 
    $extensionName = 'my_extension';
    $pluginName = 'MyCoolPlugin';
-   $pluginSignature == "myextension_mycoolplugin"
+   $pluginSignature = "myextension_mycoolplugin"
```
Comment (sfroemkenjw) [inline Documentation/ExtensionArchitecture/BestPractises/NamingConventions.rst:373]: Good example

File: Documentation/ExtensionArchitecture/BestPractises/NamingConventions.rst [hunk 18]
```diff
@@ -365,45 +370,49 @@ Example:
...
+    );
 
-   // Disable the display of layout and select_key fields for the plugin
-   $GLOBALS['TCA']['tt_content']['types']['list']['subtypes_excludelist'][$pluginSignature]
-       = 'layout,select_key,pages';
+    ExtensionManagementUtility::addToAllTCAtypes(
+        'tt_content',
+        '--div--;Configuration,pi_flexform,',
+        $pluginSignature,
+        'after:subheader',
```
Comment (sfroemkenjw) [inline Documentation/ExtensionArchitecture/BestPractises/NamingConventions.rst:408]: Works only for CType "header" as expected

File: Documentation/ExtensionArchitecture/BestPractises/NamingConventions.rst [hunk 19]
```diff
@@ -412,6 +421,8 @@ Example register and configure a non-Extbase plugin:
       settings.pageId = 42
    }
 
+.. _naming-conventions-plugin-key:
```
Comment (sfroemkenjw) [inline Documentation/ExtensionArchitecture/BestPractises/NamingConventions.rst:424]:
```suggestion
..  _naming-conventions-plugin-key:
```

File: Documentation/ExtensionArchitecture/BestPractises/NamingConventions.rst [hunk 20]
```diff
@@ -438,49 +449,57 @@ The plugin key used in :php:`registerPlugin()` and :php:`configurePlugin()`
...
-.. code-block:: typoscript
-   :caption: EXT:examples/Configuration/setup.typoscript
+..  code-block:: typoscript
+    :caption: EXT:examples/Configuration/setup.typoscript
 
-   plugin.tx_examples_htmlparser {
-      settings.pageId = 42
-   }
+    plugin.tx_examples_htmlparser {
+       settings.pageId = 42
```
Comment (sfroemkenjw) [inline Documentation/ExtensionArchitecture/BestPractises/NamingConventions.rst:501]:
```suggestion
      settings.pageId = 42
```

File: Documentation/ExtensionArchitecture/Extbase/Reference/FrontendPlugins.rst [hunk 21]
```diff
@@ -43,10 +43,16 @@ Use the following steps to add the plugin as content element:
...
 
-    #. Extension key :php:`'blog_example'` or name :php:`BlogExample`.
-    #. A unique identifier for your plugin in UpperCamelCase: :php:`'PostSingle'`
-    #. An array of allowed combinations of controllers and actions stored in an array
-    #. (Optional) an array of controller name and  action names which should not be cached
+    #.  Extension key :php:`'blog_example'` or name :php:`BlogExample`.
+    #.  A unique identifier for your plugin in UpperCamelCase: :php:`'PostSingle'`
+    #.  An array of allowed combinations of controllers and actions stored in an array
+    #.  (Optional) an array of controller name and  action names which should not be cached
+    #.  Using any value but `ExtensionUtility::PLUGIN_TYPE_CONTENT_ELEMENT` is
```
Comment (sfroemkenjw) [inline Documentation/ExtensionArchitecture/Extbase/Reference/FrontendPlugins.rst:50]: Only `list_type` will create a deprecation log entry

File: Documentation/ApiOverview/ContentElements/ContentElementsWizard.rst (hunk 2 above)
Comment (linawolf) [inline Documentation/ApiOverview/ContentElements/ContentElementsWizard.rst:82]:
```suggestion
    elements and plugins should be added with string `CType` for parameter `$type`.
```

File: Documentation/ApiOverview/ContentElements/Index.rst (hunk 4 above)
Comment (linawolf) [inline Documentation/ApiOverview/ContentElements/Index.rst:189]:
```suggestion
    Setting the second parameter to `list_type`
```

File: Documentation/ApiOverview/ContentElements/MigrationListType.rst [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+Add value `ExtensionUtility::PLUGIN_TYPE_CONTENT_ELEMENT` as fifth parameter,
+`$pluginType`, to the method :php:`ExtensionUtility::configurePlugin()`:
+
+..  literalinclude:: _Migration/_ext_localconf.php.diff
+    :caption: EXT:examples/ext_localconf.php (difference)
+
+If the fourth parameter, `$nonCacheableControllerActions` was missing you can
+set it to an empty array, the default value.
+
+It is theoretically possible that the extension author did not use this utility
```
[the remark this answers is not in the document]
Comment (linawolf) [reply Documentation/ApiOverview/ContentElements/MigrationListType.rst:53]: Yes and I have seen ppl do that. You cannot rely on they used the Utility functions. But "theoretically"

File: Documentation/ApiOverview/ContentElements/_Migration/_ext_localconf.php.diff (hunk 13 above)
Comment (linawolf) [reply Documentation/ApiOverview/ContentElements/_Migration/_ext_localconf.php.diff:12]: This is a diff, not a PHP file?

File: Documentation/CodeSnippets/FlexForms/Examples/PluginHaikuListRegistration.rst.txt (hunk 14 above)
Comment (linawolf) [reply Documentation/CodeSnippets/FlexForms/Examples/PluginHaikuListRegistration.rst.txt:23]: Actually in non-Extbase plugins this is just an arbitary string tha you can define yourself?

File: Documentation/CodeSnippets/FlexForms/Examples/PluginHaikuListRegistration.rst.txt (hunk 15 above)
Comment (linawolf) [reply Documentation/CodeSnippets/FlexForms/Examples/PluginHaikuListRegistration.rst.txt:40]: Best practice: Always keep a comma at the end of a TCA field list: Two consecutive commas are ignored but if one is missing, adding the next field will fail.

File: Documentation/CodeSnippets/FlexForms/Examples/PluginHaikuListRegistration.rst.txt (hunk 16 above)
Comment (linawolf) [inline Documentation/CodeSnippets/FlexForms/Examples/PluginHaikuListRegistration.rst.txt:42]:
```suggestion
        'after:header',
```

File: Documentation/ExtensionArchitecture/Extbase/Reference/FrontendPlugins.rst (hunk 21 above)
Comment (linawolf) [reply Documentation/ExtensionArchitecture/Extbase/Reference/FrontendPlugins.rst:50]: And the default value which is empty but defaults to list_type as well. So for all practical reasons: just use ExtensionUtility::PLUGIN_TYPE_CONTENT_ELEMENT. I don't think details beyond this are needed

File: Documentation/CodeSnippets/FlexForms/Examples/PluginHaikuListRegistration.rst.txt (hunk 14 above)
Comment (linawolf) [reply Documentation/CodeSnippets/FlexForms/Examples/PluginHaikuListRegistration.rst.txt:23]: This needs to be changed in the example extension and will automatically be updated here. If you want to make a PR, go ahead!

File: Documentation/ExtensionArchitecture/BestPractises/NamingConventions.rst [hunk 23]
```diff
@@ -365,45 +370,49 @@ Example:
...
+    );
 
-   // Disable the display of layout and select_key fields for the plugin
-   $GLOBALS['TCA']['tt_content']['types']['list']['subtypes_excludelist'][$pluginSignature]
-       = 'layout,select_key,pages';
+    ExtensionManagementUtility::addToAllTCAtypes(
+        'tt_content',
+        '--div--;Configuration,pi_flexform,',
+        $pluginSignature,
+        'after:subheader',
```
Comment (linawolf) [inline Documentation/ExtensionArchitecture/BestPractises/NamingConventions.rst:408]:
```suggestion
        'after:header',
```

File: Documentation/ApiOverview/ContentElements/MigrationListType.rst (hunk 12 above)
Comment (linawolf) [reply Documentation/ApiOverview/ContentElements/MigrationListType.rst:120]: I feel like something is missing...

File: Documentation/ApiOverview/ContentElements/MigrationListType.rst (hunk 8 above)
Comment (linawolf) [reply Documentation/ApiOverview/ContentElements/MigrationListType.rst:37]: I woul like to have the single steps linkable and that does not work with a bignum list

File: Documentation/ApiOverview/ContentElements/MigrationListType.rst (hunk 12 above)
Comment (linawolf) [reply Documentation/ApiOverview/ContentElements/MigrationListType.rst]: I added the missing content

File: Documentation/ApiOverview/ContentElements/MigrationListType.rst [hunk 24]
```diff
@@ -0,0 +1,161 @@
...
+Core-based plugin.
+
+..  _plugins-list-type-migration-extbase-flexform:
+
+2. Adjust the registration of FlexForms and additional fields
+-------------------------------------------------------------
+
+..  literalinclude:: _Migration/_tca_registration.php.diff
+    :caption: EXT:examples/Configuration/TCA/Overrides/tt_content.php (difference)
+    :linenos:
```
Comment (sfroemkenjw) [inline Documentation/ApiOverview/ContentElements/MigrationListType.rst:65]: Duplicate
```suggestion
```

File: Documentation/ApiOverview/ContentElements/MigrationListType.rst [hunk 25]
```diff
@@ -0,0 +1,161 @@
...
+    :linenos:
+
+The `CType` based plugin does not inherit the standard fields provided by the
```
Comment (sfroemkenjw) [inline Documentation/ApiOverview/ContentElements/MigrationListType.rst:68]:
```suggestion
The `CType` based plugin does not inherit the default fields provided by the
```

**Comment by linawolf** [comment]
Further fixes can be made in follow ups

**Pull request merged**